spec/models/job_proxy_dispatcher_spec.rb

```diff
@@ -172,5 +175,103 @@ module JobProxyDispatcherSpec
+            expect(x.target_class).to eq 'ContainerImage'
+          end
+          expect(jobs.count).to be > 0 # in case something unexpected goes wrong
+        end
+      end
+
+      describe "#pending_container_jobs" do
+        it "returns container jobs by provider" do
+          jobs_by_ems, = @dispatcher.pending_container_jobs
+          expect(jobs_by_ems.keys).to match_array(@container_providers.map(&:id))
+
+          expect(jobs_by_ems[@container_providers.first.id].count).to eq(1)
+          expect(jobs_by_ems[@container_providers.second.id].count).to eq(2)
+        end
+      end
+
+      describe "#active_container_scans_by_zone_and_ems" do
+        it "returns active container acans for zone" do
+          job = @jobs.find { |j| j.target_class == ContainerImage.name }
+          job.update(:dispatch_status => "active")
+          provider = ExtManagementSystem.find(job.options[:ems_id])
+          @dispatcher.instance_variable_set(:@zone, MiqServer.my_zone) # memoized during pending_jobs call
+          expect(@dispatcher.active_container_scans_by_zone_and_ems).to eq(
+            job.zone => {provider.id => 1}
+          )
+        end
+      end
+
+      describe "#dispatch_container_scan_jobs" do
+        it "dispatches jobs until reaching limit" do
+          stub_settings(:container_scanning => {:concurrent_per_ems => 1})
+          @dispatcher.dispatch_container_scan_jobs
+          expect(Job.where(:target_class => ContainerImage, :dispatch_status => "pending").count).to eq(1)
+          # 1 per ems, one ems has 1 job and the other 2
+        end
+
+        it "does not dispach if limit is already reached" do
+          stub_settings(:container_scanning => {:concurrent_per_ems => 1})
+          @dispatcher.dispatch_container_scan_jobs
+          expect(Job.where(:target_class => ContainerImage, :dispatch_status => "pending").count).to eq(1)
+          @dispatcher.dispatch_container_scan_jobs
+          expect(Job.where(:target_class => ContainerImage, :dispatch_status => "pending").count).to eq(1)
+        end
+
+        it "does not apply limit when concurrent_per_ems is 0" do
+          stub_settings(:container_scanning => {:concurrent_per_ems => 0})
+          @dispatcher.dispatch_container_scan_jobs
+          expect(Job.where(:target_class => ContainerImage, :dispatch_status => "pending").count).to eq(0)
+          # 1 per ems, one ems has 1 job and the other 2
+        end
+
+        it "does not apply limit when concurrent_per_ems is -1" do
+          stub_settings(:container_scanning => {:concurrent_per_ems => -1})
+          @dispatcher.dispatch_container_scan_jobs
+          expect(Job.where(:target_class => ContainerImage, :dispatch_status => "pending").count).to eq(0)
+          # 1 per ems, one ems has 1 job and the other 2
+        end
+      end
+    end
+
+    describe "#active_vm_scans_by_zone" do
+      it "returns active vm scans for this zone" do
+        @jobs = (@vms + @repo_vms).collect(&:scan)
```

Comment:
Same as below "Why are we scanning several vms? Does it matter? Why not just 1?"

I don't like extract setup in a test because it's slow and make it appear that all of this extra stuff is required for the test.